### DISCUSSION TITLE
Optuna contour plot fails to display with too many data points

### DISCUSSION THREAD

**Discussion by stephano41** [optuna/optuna#3830] [Q&A]
I've ran a study with over 1000 trials and there were several parameters tuned, when I tried to plot the contour plot using optuna.visualization.plot_contour, my browser to display the plot just crashes. I was wondering if there was a way to get around this and show the contour plot.

I've tried using the matplotlib version, but as there were too many data points, it became very hard to see anything on the plot:
![image](https://user-images.githubusercontent.com/78075921/181728142-9842fad3-9e1a-4a59-a4ec-a1d04325e4ef.png)

Any ideas on how to display all this data with a contour plot?

**Comment by nzw0301** [comment]
Yes, I suppose so if the number of trials is too large. One possible solution is to sub-sample a trial for visualisation. For example, extract a part of trials from the study object, then add a part of them to an empty study by using https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.add_trials.

**Comment by blazespinnaker** [comment]
It would probably be a good idea to first just try just doing a contour plot for pairs that you're interested in, one by one.  Then use zoom a lot.  You may also want to consider pruning, though it's not clear to me how that impacts the samplers.

**Reply by stephano41** [reply]
Thank you @blazespinnaker @nzw0301 for your suggestions!

Down-sampling trials could be the potential solution, however I am also unsure how to go about deciding the down sampling strategy (random or other more sophisticated means) so that I still preserve enough data to tell what is valid

I've tried exporting the plotly image as pdf and that's so far been the only thing that's managed to open the image in a vectorised form that I can keep zooming in. However the black dots are quite big still, is there any way to shrink the black dots? Happy to copy and paste the plot_contour code and play with it, but any directions as to where the size of the black dots are defined would be great!
![image](https://user-images.githubusercontent.com/78075921/183095664-62eeea91-1ab5-4987-b6e4-9f2ef7c7d494.png)

**Reply by nzw0301** [reply]
Since the plot object comes from plotly API, how about changing the style of the figure by specifying a smaller value to `fig.data[1]["marker"]["size"]`. For example, 

```python
import optuna


def objective(trial):
    x = trial.suggest_float("x", -100, 100)
    y = trial.suggest_categorical("y", [-1, 0, 1])
    return x ** 2 + y


sampler = optuna.samplers.TPESampler(seed=10)
study = optuna.create_study(sampler=sampler)
study.optimize(objective, n_trials=30)

fig = optuna.visualization.plot_contour(study, params=["x", "y"])

# make marker size smaller
fig.data[1]["marker"]["size"] = 1

fig.write_image("sample.pdf")
```

Ref: https://plotly.com/python/marker-style/

**Comment by blazespinnaker** [comment]
Try something like optuna.visualization.plot_contour(study, params=["p1", "p2"])

**Comment by stephano41** [comment]
I ended up using:
```python
figure = plot_contour(study, target=lambda t: t.values[0])

for i in range(0, len(figure.data)):
    if hasattr(figure.data[i], "marker"):
        figure.data[i]["marker"]["size"] = 0.1
```

which works!
